File: BOJ/BFS&DFS/BJ1260-2.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().split('\n');

const [cnt, ...edgeStr] = input;
const [N, M, V] = cnt.split(' ').map(Number);
const edges = edgeStr.map(v => v.split(' ').map(Number));

const graph = JSON.parse(JSON.stringify(new Array(N + 1).fill([])));
let s = new Set();
let result = [];

for(let i=0; i<M; i++) {
  const [start, end] = edges[i];
  
  graph[start].push(end);
  graph[end].push(start);
};

graph.forEach(v => {
  v.sort((a, b) => a - b);
});

function dfs(cur) {
  s.add(cur);

  result.push(cur);

  for(const next of graph[cur] || []) {
    if(s.has(next)) continue; 
    dfs(next);
  }
}

function bfs (start) {
  const queue = [];
  queue.push(start);

  while(queue.length > 0) {
    const cur = queue.shift();
    result.push(cur);

    for(const next of graph[cur] || []) {
      if(s.has(next)) continue;
      s.add(next);
      queue.push(next);
    }
  };
}

dfs(V);
console.log(result.join(' '));
s = new Set([V]);
result = [];
bfs(V);
console.log(result.join(' '));

File: BOJ/BFS&DFS/BJ1260-2.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <utility>
#include <algorithm>
#include <cstring>
using namespace std;

int N, M, V;
vector<vector<int> > graph;
queue<int> q;
bool is_visited[1001];

void dfs(int cur) {
  if(is_visited[cur]) return;
  is_visited[cur] = true;
  cout << cur << ' ';
  for(int i=0; i<graph[cur].size(); i++) {
    int next = graph[cur][i];
    if(is_visited[next]) continue;
    dfs(next);
  }
}

void bfs(int start) {
  q.push(start);
  is_visited[start]=true;
  
  while(!q.empty()) {
    int cur = q.front();
    cout << cur << ' ';
    q.pop();

    for(int i=0; i<graph[cur].size(); i++) {
      int next = graph[cur][i];
      if(is_visited[next]) continue;
      is_visited[next]=true;
      q.push(next);
    }
  }
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  cin >> N >> M >> V;

  graph.resize(N + 1);

  int start, end;
  for(int i=0; i<M; i++) {
    cin >> start >> end;
    
    graph[start].push_back(end);
    graph[end].push_back(start);
  }
  
  for(int i=0; i<N; i++) {
    sort(graph[i].begin(), graph[i].end());
  }

  dfs(V);
  memset(is_visited, 0, sizeof(is_visited));
  cout << '\n';
  bfs(V);
  return 0;
}